### PR TITLE
Add validation error to prevent checkout when there is no shipping method available

### LINKED ISSUE #8348: Local pickup: Add validation error to prevent checkout when there is no shipping method
## Describe the bug

When there is no shipping method, local pickup won't remove the pickup cost from the Cart totals. We're able to place an order even when there is no shipping method available.

## To reproduce

1. Enable local pickup with pickup price
2. Remove the shipping methods from the WC settings
3. Go to the Checkout block and select the shipping option
4. Place an order

## Expected behavior

There should be a validation error that prevents the checkout. You should not be allowed to place an order

## Screenshots

https://user-images.githubusercontent.com/11503784/215834024-fd051ffe-d3dc-4abf-8ead-f59773ca323c.mov

### PR DESCRIPTION
**Pull request by tarunvijwani**
With this PR, we're preventing checkout if no shipping methods are available. 

Fixes #8348

## Changes in the PR:
- Add the validation error in ShippingSelector to prevent checkout when no shipping rates are available.
- Add a notice to the checkout processor so that notice appears after clicking on the place order button.

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

| Before | After |
| ------ | ----- |
|image|image|

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Enable local pickup with pickup price.
2. Remove the shipping methods from the WC settings.
3. Go to the Checkout block and select the shipping option.
4. Confirm error notice is being displayed at the top.
5. Confirm you're not able to place the order.
6. Select local pickup.
7. Confirm you're able to place an order

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Add validation error to prevent checkout when there is no shipping method available

### CODE REVIEW

File: assets/js/base/components/cart-checkout/totals/shipping/index.tsx
Comment (senadir) [inline assets/js/base/components/cart-checkout/totals/shipping/index.tsx]: This doesn't seem like the right place to trigger this, it should be at the pickup/shipping selector level to ensure it only runs in Checkout and not both Cart and Checkout.

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx
Comment (senadir) [inline assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx]: Shouldn't this error trigger at the shipping selector level? ideally we would only run into this scenario when we have both local pickup and regular shipping.

**Review by senadir — CHANGES_REQUESTED** [review]
~This is not working for me :/~

This is now working for me, but it still seems to be buggy.

If you go regular shipping, place, get an error, switch to local pickup, the error goes away, switch back to regular shipping, the error surfaces again for some seconds before it disappears.

**Review by senadir — CHANGES_REQUESTED** [review]
Hey @tarunvijwani , we're going to put this on hold until @elizaan36 figures out some design next steps when it comes to validation and text.

**Comment by elizaan36** [comment]
When the user first lands on the Checkout block with no address provided, can we hide the "Shipping options" section?

Then if an address is entered that's not within a valid shipping zone, we can present a notice inline in the shipping options section. The top level notice isn't necessary in this case. Let me know what you think @tarunvijwani @senadir 

image

**Comment by senadir** [comment]
Hey @elizaan36 this will still get us to the same place of needing something to prevent the customer from placing order.

The issue here is that a shopper still has Local Pickup selected (even if not visible) and they can place the order with the error shown.

**Comment by ralucaStan** [comment]
Noting that we don't have information-type notices implemented in the project. We could display a normal error notice in the shipping section until we have the new notice types added in the blocks. Does that sound good to you @elizaan36 ?

**Comment by elizaan36** [comment]
Yep, sounds good. Based on the conversation in Slack, there should be the same pattern/message for the same error whether local pickup is enabled or not.

When pickup is not enabled, there's an error notice in the Shipping options section. 
> There are no shipping options available. Please check your shipping address.

Then if user proceeds to checkout there's a top level notice.
> Sorry, this order requires a shipping option.

**Comment by tarunvijwani** [comment]
Thank you, @senadir and @elizaan36, for the review. I have updated the code. 👍

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx
Comment (tarunvijwani) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/package-rates.tsx]: Yes, thank you for the suggestion. I have moved the code to the Shipping selector.

File: assets/js/base/context/providers/cart-checkout/checkout-processor.ts
Comment (mikejolley) [inline assets/js/base/context/providers/cart-checkout/checkout-processor.ts]: Since this is within a callback and `shippingRatesValidationError` is not used anywhere else, could we move the lookup of this error inline?

e.g. `select( VALIDATION_STORE_KEY ).getValidationError( 'shipping-rates-error' ) !== undefined`

Might be cleaner?

File: assets/js/base/context/providers/cart-checkout/checkout-processor.ts
Comment (mikejolley) [inline assets/js/base/context/providers/cart-checkout/checkout-processor.ts:171]: This needs to be localised.

File: assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx
Comment (mikejolley) [inline assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx]: This needs to be localised.

**Review by mikejolley — CHANGES_REQUESTED** [review]
Missing some localisation functions here, but it tested ok!

File: assets/js/data/utils/process-error-response.ts
Comment (tarunvijwani) [inline assets/js/data/utils/process-error-response.ts]: This change will fix the TypeScript error present in the `assets/js/base/context/providers/cart-checkout/checkout-processor.ts` file for not passing [two arguments](https://github.com/woocommerce/woocommerce-blocks/blob/8932bc96ad8588e3feda2eba89014d1d19d92646/assets/js/base/context/providers/cart-checkout/checkout-processor.ts#L292).

File: assets/js/base/context/providers/cart-checkout/checkout-processor.ts
Comment (tarunvijwani) [reply assets/js/base/context/providers/cart-checkout/checkout-processor.ts]: That's a great suggestion. I have updated the code.

File: assets/js/base/context/providers/cart-checkout/checkout-processor.ts
Comment (tarunvijwani) [reply assets/js/base/context/providers/cart-checkout/checkout-processor.ts:171]: Ah missed localizing the strings. Thank you for bringing this up. I have added the localization.

**Comment by tarunvijwani** [comment]
@mikejolley Thank you for the review. I have updated the code. 👍

**Comment by mikejolley** [comment]
@tarunvijwani this needs a rebase - ready for final review?

File: assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/rate-price.tsx
Comment (tarunvijwani) [inline assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/shared/rate-price.tsx:19]: This change will fix the TypeScript error in `assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx` for not passing the `multiple` prop.

**Comment by tarunvijwani** [comment]
> @tarunvijwani this needs a rebase - ready for final review?

@mikejolley I have fixed the conflicts. This is now ready for review. Thank you! 🙂

**Review by mikejolley — APPROVED** [review]
👍🏻

**Pull request merged**
